common/src/graph.rs

```diff
@@ -231,27 +208,49 @@ impl Graph {
+        // to the not-yet-created neighbor node.
+        let mut shorter_neighbors_of_subject = [None; 3]; // Maximum number of shorter neighbors is 3
         let mut count = 0;
-        for neighbor_side in Side::iter() {
-            if neighbor_side == parent_side
-                || !neighbor_side.adjacent_to(parent_side)
-                || !self.is_near_side(parent, neighbor_side)
+        for candidate_descender in Side::iter() {
+            if candidate_descender == side {
+                // The given node is included in the list of returned nodes.
+                shorter_neighbors_of_subject[count] = Some((side, node));
+                count += 1;
+            } else if candidate_descender.adjacent_to(side)
+                && self.is_descender(node, candidate_descender)
             {
-                continue;
+                // This branch covers shorter neighbors of the subject other than the given node.
+                // This is non-obvious, as it relies on the fact that a side is a descender of the subject
+                // exactly when it is a descender of the given node. This is not true in general, but it is true
+                // when the descender in question is adjacent to the side shared by the given node and the subject.
+                // That is what allows the `self.is_near_side(node, candidate_descender_side)` condition to behave as desired.
+
+                // We would like to return the shorter neighbor of the subject. This means that
+                // if we label the shared side A and the descender B, the path we would like to follow from the given node is AB,
+                // since A will take us to the subject, and then B will take us to its shorter neighbor. However, taking
+                // the path AB is impossible because it would require the subject to already be in the graph. Fortuantely,
+                // we can take the path BA instead because that will reach the same node, thanks to the fact that each edge
+                // is shared by 4 dodecas.
+                let shorter_neighbor_of_node = self.neighbor(node, candidate_descender).unwrap();
+                let shorter_neighbor_of_subject =
+                    self.ensure_neighbor(shorter_neighbor_of_node, side);
+                shorter_neighbors_of_subject[count] =
+                    Some((candidate_descender, shorter_neighbor_of_subject));
+                count += 1;
+            } else {
+                // The `candidate_descender_side` is not a descender of the subject, so no action is necessary.
             }
-            let x = self.nodes[&parent].neighbors[neighbor_side as usize].unwrap();
-            let neighbor = self.ensure_neighbor(x, parent_side);
-            neighbors[count] = Some((neighbor_side, neighbor));
-            count += 1;
         }
-        (0..2).filter_map(move |i| neighbors[i])
+        (0..3).filter_map(move |i| shorter_neighbors_of_subject[i])
```

Comment:
While we're revisiting this, might as well take advantage of the `[T]::into_iter` that's been introduced since it was written, just to be a bit less weird:
```suggestion
        shorter_neighbors_of_subject.into_iter().filter_map(|x| x)
```

Reply:
Thanks! That does seem nice. Clippy further simplifies `filter_map(|x| x)` to `flatten()`, which seems a bit weird to me, but that's only because I consider the concept of an `Option` being a type of iterator to be a bit weird conceptually. I'll trust the designers of Rust that this is okay.

Reply:
Think of it as a container of 0 or 1 items, just like an array is a container of N items, or a Vec is a container of an arbitrary number of items.

Reply:
Oh, yeah. I do think of it that way. It's just not the most natural way for me.